core/src/banking_stage.rs

```diff
@@ -839,9 +846,13 @@ impl BankingStage {
         if unprocessed_transaction_storage.should_not_process() {
             return;
         }
-        let ((metrics_action, decision), make_decision_time) = measure!(
-            Self::make_consume_or_forward_decision(my_pubkey, poh_recorder, slot_metrics_tracker)
-        );
+        let ((metrics_action, decision), make_decision_time) =
+            measure!(Self::make_consume_or_forward_decision(
+                my_pubkey,
+                poh_recorder,
+                slot_metrics_tracker,
+                unprocessed_transaction_storage.len(),
```

Comment:
What is the benefit of including this metric into the trace file? It doesn't seem particularly useful for a new simulate tool, and iirc we already capture this metric in our regular metric reporting.

Reply:
> as a general design policy, I'd like for the simulation to work by its own trace files

agree, this is a necessity for the tool to be useful. I'm just not sure where this will be used in the simulate tool. We have timestamps of when the packets arrive from sigverify, and when each leader bank slot starts.

Performance of how quickly BS processes the packets is going to depend on the implementation at simulation time - I'm just struggling to think how the length of the buffer plays into the simulation part, except maybe for initialization for buffer capacity?

Reply:
Okay, I think I see how it's being used in the simulation tool now (see below). But think the use leads us into a deeper discussion of "what is the goal of the simulator?".

Is the goal of the simulator:
1. replicate sequence of events as closely as possible to what happened in actual execution or
2. simulate how banking stage would perform given same load?
Potentially it is both, and I missed that.

I had previously been viewing this from goal 2's perspective and hadn't considered goal 1. With this in mind, I had imagined the simulator tool would just use timestamp differences to determine when to send packet batches across the channel during simulation, and the banking stage would receive+buffer naturally based on how its' performing.

Having a look at your wip simulator tool, it uses these `unprocessed_counts` to limit packet inflow to banking stage **based on when those packets were received/buffered during actual execution**. I think this fits into goal 1 above, since we're artificially limiting packet inflow to BS based on how BS executed in the actual leader slot.

Reply:
thanks for clarifying around the goals of this thing. :)

to tell the truth, even i was mixed with the separate goals to some extent.

as a quick recap, I was occupied with 1. when prototyping early days to validate feasibility of simulation at all outside `solana-validator`.

at later stage, I'm more leaning towards the goal 2. now that after writing this doc (https://github.com/ryoqun/solana/pull/2/commits/5b2ba785427c1d737e6a8a01667087f5e35bdeb7), i also noticed use of `unprocessed_counts` isn't aligheing with the goal 2 (noted as todos in the doc comments). So, I'll change simulation code not to use it while validating similar reproducible performance.